### ISSUE TITLE
Recognize undefined variables

### ISSUE THREAD

**Issue by altoorodevelead** [pmd/pmd#2804]
**Description:**
For the following code:
```java
class HelloWorld 
{ 
  public static void main(String args[]) 
  { 
    System.out.println(soso);
    int i = 4; 
  }       
}
```

Running PMD: 
```sh
$ ./run.sh pmd -d test.java -R rulesets/java/quickstart.xml -no-cache -f json > test.json
```
Returns the following:
```json
...
"violations": [
        {
          "beginline": 1,
          "begincolumn": 1,
          "endline": 8,
          "endcolumn": 1,
          "description": "All classes, interfaces, enums and annotations must belong to a named package",
          "rule": "NoPackage",
          "ruleset": "Code Style",
          "priority": 3,
          "externalInfoUrl": "https://pmd.github.io/pmd-6.27.0/pmd_rules_java_codestyle.html#nopackage"
        },
        {
          "beginline": 2,
          "begincolumn": 1,
          "endline": 8,
          "endcolumn": 1,
          "description": "All methods are static.  Consider using a utility class instead. Alternatively, you could add a private constructor or make the class abstract to silence this warning.",
          "rule": "UseUtilityClass",
          "ruleset": "Design",
          "priority": 3,
          "externalInfoUrl": "https://pmd.github.io/pmd-6.27.0/pmd_rules_java_design.html#useutilityclass"
        },
        {
          "beginline": 6,
          "begincolumn": 9,
          "endline": 6,
          "endcolumn": 9,
          "description": "Avoid unused local variables such as \u0027i\u0027.",
          "rule": "UnusedLocalVariable",
          "ruleset": "Best Practices",
          "priority": 3,
          "externalInfoUrl": "https://pmd.github.io/pmd-6.27.0/pmd_rules_java_bestpractices.html#unusedlocalvariable"
        }
      ]
``` 
Does a rule exist to catch **undefined** variables like _"soso"_ ?

Thanks in advance.

**Comment by senbagaraman04** [comment]
As the compiler detects the undefined ones while compiling and most of the modern cases where the IDE itself is more capable of finding this type of error. Coming to your question, there is no rule to detect the undefined rules, as per this page : [Docs Rules pmd](https://pmd.github.io/latest/pmd_rules_java.html)

**Comment by jsotuyod** [comment]
@senbagaraman04 PMD analysis requires compiled classes for best results. Therefore, it will always assume the code is valid and compiles properly.

Also, in the aim to add value to developers, we see little to gain reporting something the compiler already does.

**Comment by altoorodevelead** [comment]
Thanks for the answers. 
It makes a lot sense.
